### PR TITLE
Elixir update

### PR DESCRIPTION
**Pull request by happysalada**
###### Motivation for this change

Following a discussion from https://github.com/NixOS/nixpkgs/pull/129094

There was a breaking change in elixir 1.12.2 in the way libraries are loaded when building.
Before 1.12.2 adding a library path to ERL_LIBS was sufficient for mix to find the library
Now the library needs to be present in the build directory.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Relase notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by NinjaTrappeur** [comment]
@GrahamcOfBorg test pleroma

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 131769` at 4dd2c465 run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li><a href="https://gist.github.com/3fa36eca9705ae78bb26541c1829ce7c">pleroma</a></li>
  </ul>
</details>
<details>
  <summary>3 packages built successfully:</summary>
  <ul>
    <li>elixir (elixir_1_12)</li>
    <li>elixir_ls</li>
    <li>plausible</li>
  </ul>
</details>

Note that build failures [may predate](https://github.com/nix-community/hydra-check) this PR, and could be nondeterministic or hardware dependent.
Please exercise your independent judgement. Does something look off? Please file an [issue](https://github.com/rmcgibbo/r-rmcgibbo/issues) or reach out on IRC.

**Comment by happysalada** [comment]
I've noticed the failure from the bot to build pleroma.
I've added another fix that I think should work.
I have no way of testing pleroma on my computer. I would be happy to make more fixes if anybody has time to test and upload the logs.

File: pkgs/development/beam-modules/build-mix.nix [hunk 1]
```diff
@@ -60,15 +80,16 @@ let
 
       # Some packages like db_connection will use _build/shared instead of
       # honoring the $MIX_ENV variable.
-      for reldir in _build/{$MIX_ENV,shared}/lib/${name}/{src,ebin,priv,include} ; do
+      for reldir in _build/{$MIX_ENV,shared}/lib/${name}/{src,ebin,priv,include};
+      do
```
Comment (SuperSandro2000) [inline pkgs/development/beam-modules/build-mix.nix:84]:
```suggestion
      for reldir in _build/{$MIX_ENV,shared}/lib/${name}/{src,ebin,priv,include}; do
```

File: pkgs/development/beam-modules/mix-release.nix [hunk 2]
```diff
@@ -58,6 +58,20 @@ stdenv.mkDerivation (overridable // {
...
+        # Strip version number for directory name if it exists, so naming of
+        # all libs matches what mix's expectation.
+        dest=$(basename $dir | cut -d '-' -f1)
+        echo "Linking $dir to _build/$MIX_ENV/lib/$dest"
+        # Symlink libs to _build so that mix can find them when compiling.
+        # This is what allows mix to compile the package without searching
+        # for dependencies over the network.
+        ln -s $dir _build/$MIX_ENV/lib/$dest
+      done
+    done <<< "$ERL_LIBS:"
```
Comment (SuperSandro2000) [inline pkgs/development/beam-modules/mix-release.nix:73]: Not sure if we want to do something about this code duplication.

File: pkgs/development/beam-modules/build-mix.nix (hunk 1 above)
Comment (happysalada) [reply pkgs/development/beam-modules/build-mix.nix:84]: I messed up this commit. This should definitely not have been there

File: pkgs/development/beam-modules/mix-release.nix (hunk 2 above)
Comment (happysalada) [reply pkgs/development/beam-modules/mix-release.nix:73]: Thank you for reviewing!

You're right here. I'm not sure how to abstract in a shell function and share with different derivations.

Comment (SuperSandro2000) [reply pkgs/development/beam-modules/mix-release.nix:73]: We could make a hook which is run in both files.

**Review by picnoir — APPROVED** [review]
Just tested the pleroma build & test locally, it's working great.

LGTM, Good job @happysalada <3

**Comment by NinjaTrappeur** [comment]
For the sake of completeness:

@GrahamcOfBorg

**Comment by NinjaTrappeur** [comment]
For the sake of completeness:

@GrahamcOfBorg test pleroma

File: pkgs/development/beam-modules/mix-release.nix (hunk 2 above)
Comment (happysalada) [reply pkgs/development/beam-modules/mix-release.nix:73]: I think you're right here, I don't understand yet how hooks work really well. I'm going to have a go at it soon.

**Comment by happysalada** [comment]
@NinjaTrappeur the code comes from @ydlr, he is the one to thank for on this one!

Thanks a lot for testing this!

I'm going to have a go at the hook thing before trying to merge. Almost there.

**Comment by happysalada** [comment]
@SuperSandro2000 this is my try at abstracting away the common parts.
I couldn't go through a setup hook as the semantics are not the same.

@NinjaTrappeur could you test one more time when you have a moment. This should be the final test.

**Comment by turion** [comment]
> @NinjaTrappeur the code comes from @ydlr, he is the one to thank for on this one!

You could use https://github.blog/2018-01-29-commit-together-with-co-authors/ to amend the commit

**Review by SuperSandro2000 — APPROVED** [review]

**Comment by ydlr** [comment]
> You could use https://github.blog/2018-01-29-commit-together-with-co-authors/ to amend the commit

Completely unnecessary. Please don't go through the trouble.

**Comment by ydlr** [comment]
Result of `nixpkgs-review pr 131769` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>6 packages built:</summary>
  <ul>
    <li>elixir (elixir_1_12)</li>
    <li>elixir_ls</li>
    <li>mix2nix</li>
    <li>plausible</li>
    <li>pleroma</li>
    <li>rabbitmq-server</li>
  </ul>
</details>

**Comment by happysalada** [comment]
@turion thank you for the info, I will definitely remember this for next time! There are a lot of people who are not credited in the nixpkgs elixir code base.

@SuperSandro2000 Thank you for the review!

@ydlr thank you for the review, the original code and sorry for having butchered your formatting 😄

**Pull request merged**